testsuite/forge/src/runner.rs

```diff
@@ -0,0 +1,262 @@
+        Format::Pretty
+    }
+}
+
+pub fn forge_main<F: Factory>(tests: ForgeConfig<'_>, factory: F) -> Result<()> {
+    let options = Options::from_args();
+    let forge = Forge::new(&options, tests, factory);
+
+    if options.list {
+        forge.list()?;
+
+        return Ok(());
+    }
+
+    forge.run()
+}
+
+pub struct ForgeConfig<'cfg> {
+    pub public_usage_tests: &'cfg [&'cfg dyn PublicUsageTest],
+    pub admin_tests: &'cfg [&'cfg dyn AdminTest],
+}
+
+impl<'cfg> ForgeConfig<'cfg> {
+    pub fn number_of_tests(&self) -> usize {
+        self.public_usage_tests.len() + self.admin_tests.len()
+    }
+
+    pub fn all_tests(&self) -> impl Iterator<Item = &'cfg dyn Test> + 'cfg {
+        self.public_usage_tests
+            .iter()
+            .map(|t| t as &dyn Test)
+            .chain(self.admin_tests.iter().map(|t| t as &dyn Test))
+    }
+}
+
+pub struct Forge<'cfg, F> {
+    options: &'cfg Options,
+    tests: ForgeConfig<'cfg>,
+    factory: F,
+}
+
+impl<'cfg, F: Factory> Forge<'cfg, F> {
+    pub fn new(options: &'cfg Options, tests: ForgeConfig<'cfg>, factory: F) -> Self {
+        Self {
+            options,
+            tests,
+            factory,
+        }
+    }
+
+    pub fn list(&self) -> Result<()> {
+        for test in self.tests.all_tests() {
+            println!("{}: test", test.name());
+        }
+
+        if self.options.format == Format::Pretty {
+            println!();
+            println!("{} tests", self.tests.all_tests().count());
+        }
+
+        Ok(())
+    }
+
+    pub fn run(&self) -> Result<()> {
+        let mut summary = TestSummary::new(self.tests.number_of_tests(), 0);
+        summary.write_starting_msg()?;
+
+        let mut rng = ::rand::rngs::StdRng::from_seed([0; 32]);
+        let mut swarm = self.factory.launch_swarm();
+
+        // Run PublicUsageTests
+        for test in self.tests.public_usage_tests {
+            let mut public_ctx =
+                PublicUsageContext::new(CoreContext::from_rng(&mut rng), swarm.public_info());
+            let result = run_test(|| test.run(&mut public_ctx));
+            summary.handle_result(test.name().to_owned(), result)?;
+        }
+
+        // Run AdminTests
+        for test in self.tests.admin_tests {
+            let mut admin_ctx =
+                AdminContext::new(CoreContext::from_rng(&mut rng), swarm.admin_info());
+            let result = run_test(|| test.run(&mut admin_ctx));
+            summary.handle_result(test.name().to_owned(), result)?;
+        }
+
+        summary.write_summary()?;
+
+        if !summary.success() {
+            process::exit(101);
+        }
+
+        Ok(())
+    }
+}
+
+enum TestResult {
+    Ok,
+    Failed,
+    FailedWithMsg(String),
+}
+
+fn run_test<F: FnOnce() -> Result<()>>(f: F) -> TestResult {
+    match ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(f)) {
+        Ok(Ok(())) => TestResult::Ok,
+        Ok(Err(e)) => TestResult::FailedWithMsg(format!("{:?}", e)),
```

Comment:
Given how common getting the debug into a string is, I really wish that it was just a function built into rust...